x-pack/plugins/observability_solution/dataset_quality/server/services/data_telemetry/data_telemetry_service.ts

```diff
@@ -0,0 +1,181 @@
+
+import {
+  BREATHE_DELAY_MEDIUM,
+  BREATHE_DELAY_SHORT,
+  NON_LOG_SIGNALS,
+  EXCLUDE_ELASTIC_LOGS,
+  MAX_STREAMS_TO_REPORT,
+  STARTUP_DELAY,
+  TELEMETRY_INTERVAL,
+  LOGS_DATASET_INDEX_PATTERNS,
+} from './constants';
+import {
+  getAllIndices,
+  addMappingsToIndices,
+  addNamespace,
+  groupStatsByPatternName,
+  getIndexBasicStats,
+  indexStatsToTelemetryEvents,
+  getIndexFieldStats,
+} from './helpers';
+
+import { DataTelemetryEvent } from './types';
+
+export class DataTelemetryService {
+  private readonly logger: Logger;
+  private readonly stop$ = new Subject<void>();
+
+  private telemetryStart?: TelemetryPluginStart;
+
+  // @ts-ignore: Unused variable
+  private analytics?: AnalyticsServiceSetup;
+
+  // @ts-ignore: Unused variable
+  private isInProgress = false;
+
+  private isOptedIn?: boolean = true; // Assume true until the first check
+  private esClient?: ElasticsearchClient;
+
+  constructor(logger: Logger) {
+    this.logger = logger;
+  }
+
+  public setup(analytics: AnalyticsServiceSetup) {
+    this.analytics = analytics;
+  }
+
+  public async start(telemetryStart?: TelemetryPluginStart, core?: CoreStart) {
+    this.telemetryStart = telemetryStart;
+    this.esClient = core?.elasticsearch.client.asInternalUser;
+
+    this.logger.debug(`[Logs Data Telemetry] Starting the service`);
+    timer(STARTUP_DELAY, TELEMETRY_INTERVAL)
+      .pipe(
+        takeUntil(this.stop$),
+        tap(() => (this.isInProgress = true)),
+        switchMap(() => from(this.isTelemetryOptedIn())),
+        tap((isOptedIn) => {
+          if (!isOptedIn) {
+            this.logTelemetryNotOptedIn();
+            this.isInProgress = false;
+          }
+        }),
+        filter((isOptedIn) => isOptedIn),
+        exhaustMap(() => this.collectAndSend()),
+        tap(() => (this.isInProgress = false))
+      )
+      .subscribe();
+  }
+
+  public stop() {
+    this.stop$.next();
+  }
+
+  public async isTelemetryOptedIn() {
+    this.isOptedIn = await this.telemetryStart?.getIsOptedIn();
+    return this.isOptedIn === true;
+  }
+
+  private collectAndSend() {
+    // Gather data streams and indices related to each stream of log
+    if (this.esClient) {
+      return getAllIndices({
+        esClient: this.esClient,
+        logsIndexPatterns: LOGS_DATASET_INDEX_PATTERNS,
+        excludeStreamsStartingWith: [...NON_LOG_SIGNALS, ...EXCLUDE_ELASTIC_LOGS],
+        breatheDelay: BREATHE_DELAY_MEDIUM,
+      }).pipe(
+        switchMap((dataStreamsAndIndicesInfo) => {
+          if (dataStreamsAndIndicesInfo.length > MAX_STREAMS_TO_REPORT) {
+            this.logger.debug(
+              `[Logs Data Telemetry] Number of data streams exceeds ${MAX_STREAMS_TO_REPORT}. Skipping telemetry collection.`
+            );
+            return EMPTY;
+          }
+          return of(dataStreamsAndIndicesInfo);
+        }),
+        delay(BREATHE_DELAY_SHORT),
+        switchMap((dataStreamsAndIndicesInfo) => {
```

Comment:
It looks like the processing steps below here are executed on a per-datastream basis, is that right? Even though this is just about calls between Elasticsearch and Kibana server, I think this will cause a lot of churn for bigger clusters as thousands or requests would need to go back and forth. Could we change this to fetch this information in a bulk fashion instead (maybe batched somehow)?

Reply:
I've changed the implementation to fetch _mappings and _stats at once.